source/azure_iot_hub_client.c

```diff
@@ -626,3 +724,503 @@ AzureIoTHubClientResult_t AzureIoTHubClient_SendTelemetry( AzureIoTHubClient_t *
+
+AzureIoTHubClientResult_t AzureIoTHubClient_SendDeviceTwinReported( AzureIoTHubClient_t * pxAzureIoTHubClient,
+                                                                    const uint8_t * pucReportedPayload,
+                                                                    uint32_t ulReportedPayloadLength,
+                                                                    uint32_t * pulRequestId )
+{
+    AzureIoTMQTTResult_t xMQTTResult;
+    AzureIoTHubClientResult_t xReturn;
+    AzureIoTMQTTPublishInfo_t xMQTTPublishInfo = { 0 };
+    uint8_t ucRequestID[ 10 ];
+    size_t xTopicLength;
+    az_result xCoreResult;
+    az_span xRequestID = az_span_create( ucRequestID, sizeof( ucRequestID ) );
+
+    if( ( pxAzureIoTHubClient == NULL ) ||
+        ( pucReportedPayload == NULL ) || ( ulReportedPayloadLength == 0 ) )
+    {
+        AZLogError( ( "Failed to send reported property: Invalid argument" ) );
+        xReturn = eAzureIoTHubClientInvalidArgument;
+    }
+    else if( pxAzureIoTHubClient->_internal.xReceiveContext[ azureiothubRECEIVE_CONTEXT_INDEX_TWIN ]._internal.usState !=
+             azureiothubTOPIC_SUBSCRIBE_STATE_SUBACK )
+    {
+        AZLogError( ( "Failed to send reported property: twin topic not subscribed" ) );
+        xReturn = eAzureIoTHubClientTopicNotSubscribed;
+    }
+    else
+    {
+        if( ( xReturn = prvRequestIdGet( pxAzureIoTHubClient, xRequestID,
+                                         true, pulRequestId, &xRequestID ) ) != eAzureIoTHubClientSuccess )
+        {
+            AZLogError( ( "Failed to get request id" ) );
+        }
+        else if( az_result_failed( xCoreResult =
+                                       az_iot_hub_client_twin_patch_get_publish_topic( &pxAzureIoTHubClient->_internal.xAzureIoTHubClientCore,
+                                                                                       xRequestID,
+                                                                                       ( char * ) pxAzureIoTHubClient->_internal.pucWorkingBuffer,
+                                                                                       pxAzureIoTHubClient->_internal.ulWorkingBufferLength,
+                                                                                       &xTopicLength ) ) )
+        {
+            AZLogError( ( "Failed to get twin patch topic: 0x%08x", xCoreResult ) );
+            xReturn = eAzureIoTHubClientFailed;
+        }
+        else
+        {
+            xMQTTPublishInfo.xQOS = eAzureIoTMQTTQoS0;
+            xMQTTPublishInfo.pcTopicName = pxAzureIoTHubClient->_internal.pucWorkingBuffer;
+            xMQTTPublishInfo.usTopicNameLength = ( uint16_t ) xTopicLength;
+            xMQTTPublishInfo.pvPayload = ( const void * ) pucReportedPayload;
+            xMQTTPublishInfo.xPayloadLength = ulReportedPayloadLength;
+
+            if( ( xMQTTResult = AzureIoTMQTT_Publish( &( pxAzureIoTHubClient->_internal.xMQTTContext ),
```

Comment:
Why are we using 0 for `usPacketId` here but in telemetry send we explicitly get a packetID?  Is it because of the Qos1 needing but not Qos0 (which makes sense as there's nothing to ACK, but I still want to at least double check my knowledge here).

Reply:
Yup, method response, reported property and get twin are QoS0, so no need of packet identifier.

---
In reply to: [612750989](https://github.com/Azure/azure-iot-middleware-freertos/pull/75#discussion_r612750989) [](ancestors = 612750989)